File: src/spektrafilm/runtime/process.py
"""Runtime process entry points."""

from __future__ import annotations

from spektrafilm.runtime.params_schema import RuntimePhotoParams
from spektrafilm.runtime.pipeline import SimulationPipeline
from spektrafilm.utils.preview import resize_for_preview
from spektrafilm.runtime.params_builder import (
    digest_params,
    init_params,
)

class Simulator:
    """User-facing wrapper around the runtime simulation pipeline.
    The params passed to the constructor should be static and not be changed.
    They can be refreshed with update_params or soft_update, which delegate to the internal pipeline.
    """

    def __init__(self, params: RuntimePhotoParams):
        self._pipeline = SimulationPipeline(params) # should stay private

    def process(self, image):
        """Process the input image through the simulation pipeline and return the final result."""
        return self._pipeline.process(image)

    def update_params(self, params):
        """Update the parameters of the simulation pipeline."""
        self._pipeline.update(params)

    def soft_update(self, **kwargs):
        """Soft update parameters by only changing the provided fields, keeping the rest unchanged.
        only selected safe parameters can be updated with this method
        """
        self._pipeline.soft_update(**kwargs)

    def get_timings(self):
        """Get the timings of the different stages of the simulation pipeline."""
        return self._pipeline.get_timings()

    def get_total_elapsed_time(self):
        """Get the total wall-clock time of the last process call."""
        return self._pipeline.get_total_elapsed_time()

    def format_timings(self):
        """Format the last recorded timings for display."""
        return self._pipeline.format_timings()

    def print_timings(self):
        """Print the formatted timings of the last process call."""
        self._pipeline.print_timings()


######################################################################################
# Convenience functions for single-call simulation without needing to instantiate the Simulator class.

def simulate(image, params: RuntimePhotoParams,
             digest_params_first: bool = True,
             print_timings: bool = False):
    """Convenience function to run the simulation pipeline with a single call.
    The simulator needs digested parameters to run. By default they are digested on the fly.
    If you already have digested parameters or want to digest them yourself, set digest_params_first=False.
    """
    if digest_params_first:
        params = digest_params(params)
    simulator = Simulator(params)
    result = simulator.process(image)
    if print_timings:
        simulator.print_timings()
    return result


def simulate_preview(image, params: RuntimePhotoParams,
                     digest_params_first: bool = True,
                     print_timings: bool = False):
    """Convenience function to run the simulation pipeline with a single call.
    The simulator needs digested parameters to run. By default they are digested on the fly.
    If you already have digested parameters or want to digest them yourself, set digest_params_first=False.
    """
    max_size = params.settings.preview_max_size
    result = simulate(resize_for_preview(image, max_size), params,
                      digest_params_first=digest_params_first,
                      print_timings=print_timings)
    return result


#######################################################################################################
# Legacy for ART, to be removed in the future when the old API is fully deprecated.

class AgXPhoto(Simulator):
    def __init__(self, params: RuntimePhotoParams):
        digested_params = digest_params(params)
        super().__init__(digested_params)

# photo_params is init_params
def photo_params(film_profile, print_profile) -> RuntimePhotoParams:
    """Legacy helper to build a RuntimePhotoParams with default film and print profiles.
    Build a runtime parameter object.
    It needs to be digested with digest_params before being used in the runtime pipeline.
    film_profile - label string for the film profile to use, e.g. "kodak_portra_400
    print_profile - label string for the print profile to use, e.g. "kodak_portra_endura"
    """
    params = init_params(film_profile=film_profile, print_profile=print_profile)
    params.io.full_image = True # legacy compatibility, has no effect
    params.io.preview_resize_factor = 1.0 # legacy compatibility, has no effect
    return params

__all__ = [
    "RuntimePhotoParams",
    "Simulator",
    "simulate",
    "simulate_preview",
    "AgXPhoto", # legacy for ART
    "photo_params", # legacy for ART
]

File: spectralfilm-rs/spectralfilm-core/src/runtime/process.rs
//! User-facing process convenience helpers.

use crate::runtime::params::{RuntimeParamsPatch, RuntimePhotoParams};
use crate::runtime::params_builder::digest_params;
use crate::runtime::pipeline::{SimulationError, SimulationOutput, SimulationPipeline};

pub struct Simulator {
    params: RuntimePhotoParams,
    pipeline: SimulationPipeline,
}

impl Simulator {
    pub fn new(params: RuntimePhotoParams) -> Self {
        Self { pipeline: SimulationPipeline::new(params.clone()), params }
    }

    pub fn from_json(json: &str, digest: bool) -> Result<Self, serde_json::Error> {
        let params = RuntimePhotoParams::from_json(json)?;
        let params = if digest { digest_params(params) } else { params };
        Ok(Self::new(params))
    }

    pub fn process(&mut self, image: &[f64], width: usize, height: usize) -> Result<SimulationOutput, SimulationError> {
        self.pipeline.process(image, width, height)
    }

    pub fn update_params(&mut self, params: RuntimePhotoParams) {
        self.pipeline.update(params.clone());
        self.params = params;
    }

    pub fn update_params_json(&mut self, json: &str, digest: bool) -> Result<(), serde_json::Error> {
        let params = RuntimePhotoParams::from_json(json)?;
        let params = if digest { digest_params(params) } else { params };
        self.update_params(params);
        Ok(())
    }

    pub fn soft_update(&mut self, patch: RuntimeParamsPatch, digest: bool) {
        let mut params = self.params.clone();
        params.apply_patch(patch);
        let params = if digest { digest_params(params) } else { params };
        self.update_params(params);
    }

    pub fn soft_update_json(&mut self, json: &str, digest: bool) -> Result<(), serde_json::Error> {
        let patch = RuntimeParamsPatch::from_json(json)?;
        self.soft_update(patch, digest);
        Ok(())
    }

    pub fn params(&self) -> &RuntimePhotoParams { &self.params }

    pub fn params_json(&self) -> Result<String, serde_json::Error> { self.params.to_json() }

    pub fn format_timings(&self) -> String { self.pipeline.format_timings() }
}

pub fn simulate(image: &[f64], width: usize, height: usize, params: RuntimePhotoParams, digest: bool) -> Result<SimulationOutput, SimulationError> {
    let params = if digest { digest_params(params) } else { params };
    let mut simulator = Simulator::new(params);
    simulator.process(image, width, height)
}
